### PR TITLE
Resolve tables/schemas case-insensitively in JDBC connectors

### LINKED ISSUE #3470: MYSQL Connector does not identifies Upper Case Database Name and Table Name
For reference please refer #2863 
My MYSQL Server is running on Centos.
I am using presto to query my MYSQL database using MYSQL connector where my catalog name is **mysql**.
## Scenario 1 - DataBase name and Table name is combination of upper case and lower case letters

**Database Name** - TestDB
**Table Names** - EmployeeDetails, EmployeeTable
**Query 1** - show schemas from mysql;
**Output** - 
      Schema       

 information_schema 
 performance_schema 
 testdb  
(3 rows)

Query 20150818_064410_00003_837eu, FINISHED, 1 node
Splits: 2 total, 2 done (100.00%)
0:00 [3 rows, 61B] [25 rows/s, 524B/s]

**Query 2** - show tables from mysql.testdb;
**Output** -
 Table 

(0 rows)

Query 20150818_064532_00004_837eu, FINISHED, 1 node
Splits: 2 total, 2 done (100.00%)
0:00 [0 rows, 0B] [0 rows/s, 0B/s]

In this case presto is not able to Fetch the table names which are present in database **TestDB**.

The Mysql output
mysql> show tables from TestDB;
+------------------+
| Tables_in_TestDB |
+------------------+
| EmployeeDetails  |
| EmployeeTable    |
+------------------+
2 rows in set (0.00 sec)
## Scenario 2 - DataBase name is in lower case , Table name is combination of upper case and lower case letters

**Database Name** - lowercasedb
**Table Names** - TableOne, TableTwo
**Query 1** - show schemas from mysql;
**Output** - 
       Schema       

 information_schema 
 lowercasedb
 performance_schema 
 testdb  
(4 rows)

Query 20150818_065347_00005_837eu, FINISHED, 1 node
Splits: 2 total, 2 done (100.00%)
0:00 [4 rows, 77B] [27 rows/s, 522B/s]

**Query 2** - show tables from mysql.lowercasedb;
**Output** -
  Table   

 tableone 
 tabletwo 
(2 rows)

Query 20150818_065432_00006_837eu, FINISHED, 1 node
Splits: 2 total, 2 done (100.00%)
0:00 [2 rows, 66B] [15 rows/s, 505B/s]

**Query 3** - select \* from mysql.lowercasedb.tableone;
**Output** -
Query 20150818_065535_00007_837eu failed: Table mysql.lowercasedb.tableone does not exist

In this scenario presto is able to fetch the table names but when I am accessing the table the its giving me an error as shown above.

The Mysql output
mysql> select \* from lowercasedb.TableOne;
+-----------+-----------+
| ColumnOne | ColumnTwo |
+-----------+-----------+
|         1 | Row 1     |
|         2 | Row 2     |
+-----------+-----------+
2 rows in set (0.00 sec)
## Scenario 3 - DataBase name and Table name is in lower case letters

**Database Name** - lowercasedb
**Table Names** - tableone, tabletwo
**Query 1** - show schemas from mysql;
**Output** - 
       Schema       

 information_schema 
 lowercasedb  
 lowercasetabledb  
 performance_schema 
 testdb  
(5 rows)

Query 20150818_070234_00008_837eu, FINISHED, 1 node
Splits: 2 total, 2 done (100.00%)
0:00 [5 rows, 98B] [30 rows/s, 597B/s]

**Query 2** - show tables from mysql.lowercasetabledb;
**Output** -
  Table   

 tableone 
 tabletwo 
(2 rows)

Query 20150818_070253_00009_837eu, FINISHED, 1 node
Splits: 2 total, 2 done (100.00%)
0:00 [2 rows, 76B] [17 rows/s, 652B/s]

**Query 3** - select \* from mysql.lowercasetabledb.tableone;
**Output** -
 columnone | columntwo 
-----------+-----------
         1 | Row 1  
         2 | Row 2  
(2 rows)

Query 20150818_070319_00010_837eu, FINISHED, 1 node
Splits: 2 total, 2 done (100.00%)
0:00 [2 rows, 0B] [8 rows/s, 0B/s]

In this scenario I am able to access the tables in the database.

### PR DESCRIPTION
**Pull request by kewang1024**
Fixes #3470
Relates to #2863
Issues #12657

Changes ported from prestosql/presto#614

### CODE REVIEW

**Comment by rongrong** [comment]
Please add reference to the issue to resolves, as well as the PRs from prestosql. You might want to break the second commit into multiple.

**Comment by tooptoop4** [comment]
Ship it!

**Comment by kewang1024** [comment]
Fixes prestodb/presto#3470
Relates to prestodb/presto#2863

Changes ported from prestosql/presto#614

**Comment by kewang1024** [comment]
> Please add reference to the issue to resolves, as well as the PRs from prestosql. You might want to break the second commit into multiple.

What if in the second commit, the code is actually related to each other, latter code depends on formal function interface change. In this case, do I still break them into different commits? If I do, then intermediate commit may end up in compile error.

File: presto-testing-docker/src/main/java/com/facebook/presto/testing/docker/DockerContainer.java [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+import static io.airlift.testing.Closeables.closeAllSuppress;
+import static java.lang.String.format;
+import static java.time.temporal.ChronoUnit.MINUTES;
+import static java.time.temporal.ChronoUnit.SECONDS;
+import static java.util.Objects.requireNonNull;
+
+public final class DockerContainer
+        implements Closeable
+{
+    private static final Logger LOG = Logger.get(DockerContainer.class);
```
Comment (highker) [inline presto-testing-docker/src/main/java/com/facebook/presto/testing/docker/DockerContainer.java]: We usually use lower case for `logger` though it is a constant.

**Comment by tooptoop4** [comment]
@rongrong can this be merged?

**Comment by rongrong** [comment]
@tooptoop4 Sorry got distracted in other work. I'll finish reviewing this today.

**Comment by tooptoop4** [comment]
@kewang1024 gentle ping

**Comment by kewang1024** [comment]
> @kewang1024 gentle ping

was solving the dependent issues, will merge this PR this week.

**Comment by rongrong** [comment]
The `String -> Optional<String>` change should be before the other changes. Also please address previous comments.

**Comment by kewang1024** [comment]
> The `String -> Optional<String>` change should be before the other changes. Also please address previous comments.

ok, will do!

**Comment by kewang1024** [comment]
@rongrong all issuses addressed, can you help review the code? thanks!

File: presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/BaseJdbcClient.java [hunk 2]
```diff
@@ -633,11 +633,17 @@ protected String quoted(String catalog, String schema, String table)
...
-    protected static String escapeNamePattern(String name, String escape)
+    protected static Optional<String> escapeNamePattern(Optional<String> name, Optional<String> escape)
     {
-        if ((name == null) || (escape == null)) {
+        if (!name.isPresent() || !escape.isPresent()) {
             return name;
         }
+        return Optional.of(escapeNamePattern(name.get(), escape.get()));
+    }
+    private static String escapeNamePattern(String name, String escape)
```
Comment (rongrong) [inline presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/BaseJdbcClient.java]: Add an empty line above.

Comment (kewang1024) [reply presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/BaseJdbcClient.java]: > Add an empty line above.

Got it, thanks!

**Comment by tooptoop4** [comment]
@rongrong gentle ping

**Pull request merged**